### PR TITLE
LPS-200600 Generate a new oAuth2 token if the stored is expired

### PR DESCRIPTION
**Pull request by kevenleone**
Tickets:

https://liferay.atlassian.net/browse/LPS-200429
* image
https://liferay.atlassian.net/browse/LPS-200600

I have also updated some interfaces, hope not be a problem 🤔 

Slack thread: https://liferay.slack.com/archives/CNBG06JS3/p1698856156231199

Thanks

### CODE REVIEW

**Review by rotty3000 — APPROVED** [review]
LGTM

**Comment by kevenleone** [comment]
ci:test:relevant

File: modules/apps/oauth2-provider/oauth2-provider-web/src/main/resources/META-INF/resources/js/OAuth2Client.ts [hunk 1]
```diff
@@ -94,7 +105,12 @@ class OAuth2Client {
 					tokenResponse.then((response) =>
 						Liferay.Util.SessionStorage.setItem(
 							sessionKey,
-							JSON.stringify(response),
+							JSON.stringify({
+								...response,
+								expires_after_ms:
```
Comment (bryceosterhaus) [inline modules/apps/oauth2-provider/oauth2-provider-web/src/main/resources/META-INF/resources/js/OAuth2Client.ts:110]: Why are we expiring after 1s?

File: modules/apps/oauth2-provider/oauth2-provider-web/src/main/resources/META-INF/resources/js/OAuth2Client.ts [hunk 2]
```diff
@@ -165,9 +182,17 @@ class OAuth2Client {
 			);
 
 			if (cachedTokenData !== null && cachedTokenData !== undefined) {
-				resolve(JSON.parse(cachedTokenData));
+				const cachedToken = JSON.parse(
+					cachedTokenData
+				) as IOAuth2ClientTokenResponse & {
+					expires_after_ms: number;
+				};
```
Comment (bryceosterhaus) [inline modules/apps/oauth2-provider/oauth2-provider-web/src/main/resources/META-INF/resources/js/OAuth2Client.ts:189]: Can we move this type to a declared interface above?

File: modules/apps/oauth2-provider/oauth2-provider-web/src/main/resources/META-INF/resources/js/OAuth2Client.ts (hunk 1 above)
Comment (kevenleone) [reply modules/apps/oauth2-provider/oauth2-provider-web/src/main/resources/META-INF/resources/js/OAuth2Client.ts:110]: Is not that, the response comes 

600

So I need to convert to milliseconds 

600 * 1000 ( 600.000 ) milliseconds + current time in milliseconds

File: modules/apps/oauth2-provider/oauth2-provider-web/src/main/resources/META-INF/resources/js/OAuth2Client.ts (hunk 2 above)
Comment (kevenleone) [reply modules/apps/oauth2-provider/oauth2-provider-web/src/main/resources/META-INF/resources/js/OAuth2Client.ts:189]: Sure

**Comment by kevenleone** [comment]
ci:test:sf

**Comment by kevenleone** [comment]
ci:stop

**Comment by kevenleone** [comment]
ci:test:sf

File: modules/apps/oauth2-provider/oauth2-provider-web/src/main/resources/META-INF/resources/js/OAuth2Client.ts (hunk 2 above)
Comment (kevenleone) [reply modules/apps/oauth2-provider/oauth2-provider-web/src/main/resources/META-INF/resources/js/OAuth2Client.ts]: changed

File: modules/apps/oauth2-provider/oauth2-provider-web/src/main/resources/META-INF/resources/js/OAuth2Client.ts (hunk 1 above)
Comment (bryceosterhaus) [reply modules/apps/oauth2-provider/oauth2-provider-web/src/main/resources/META-INF/resources/js/OAuth2Client.ts:111]: ah, my bad, I didn't see that we get expires_in from the response. I thought you were just arbitrarily using 1000

**Review by bryceosterhaus — APPROVED** [review]
LGTM, feel free to send to the app sec team.https://github.com/liferay-appsec/liferay-portal

**Comment by kevenleone** [comment]
resent: https://github.com/liferay-appsec/liferay-portal/pull/1483

**Pull request closed without merging**